### DISCUSSION TITLE
Broken router rules with patch: keyword

### DISCUSSION THREAD

**Discussion by tizfra** [GeekFunkLabs/fluidpatcher#90] [Q&A] [answered]
Hi,
with a older version of squishbox (0.7.15) I've wrote some router rules that helps me to navigate thru patches with button on my alesis vortex wireless 2.
Now, with the last version of software (0.8.2) that rules doesn't work anymore.
The rules are:
`- {type: cc, par1: 44, par2: 1-10=0, patch: select}`
`- {type: cc, par1: 44, par2: 11-20=7, patch: select}`
`- {type: cc, par1: 44, par2: 21-30=8, patch: select}`
`- {type: cc, par1: 21, par2: 1-127, patch: -1}`
`- {type: cc, par1: 22, par2: 1-127, patch: 1}`

@albedozero can you help me?
Thank you in advice

**Comment by albedozero** [comment]
Greetings,

Sorry about that  - I did tweak patch: rules a bit back in v0.7.13 (https://github.com/GeekFunkLabs/fluidpatcher/commit/e0b78f2996ec42f0e716fc476e904aac51a85683) (not sure why it would have still worked for you in 0.7.15, but oh well). TLDR you put a plus or minus *after* the number to increment or decrement. I knew this might break things for some people, but now you can have a rule that selects a specific patch by number, or even name, in addition to the old functionality.

**Reply by tizfra** [reply]
Thanks for the quick reply,
I've changed my router rules, but they don't work.
`- {type: cc, par1: 44, par2: 1-10=0, patch: select}`
`- {type: cc, par1: 44, par2: 11-20=7, patch: select}`
`- {type: cc, par1: 44, par2: 21-30=8, patch: select}`
`- {type: cc, par1: 21, par2: 1-127, patch: 1 -}`
`- {type: cc, par1: 22, par2: 1-127, patch: 1 +}`

Can you make me an example?
Thanks

**Reply by albedozero** [reply]
omit the space between +/-
```yaml
- {type: cc, par1: 44, par2: 1-10=0, patch: select}
- {type: cc, par1: 44, par2: 11-20=7, patch: select}
- {type: cc, par1: 44, par2: 21-30=8, patch: select}
- {type: cc, par1: 21, par2: 1-127, patch: 1-}
- {type: cc, par1: 22, par2: 1-127, patch: 1+}
```

**Reply by tizfra** [reply]
I've done some tests and ive found that the last 2 rules works if I add `chan: 1` at the begin.
The other 3 I don't know why if I add `chan: 1`at the begin I don't see any value when the button is pressed in midi monitor, I see only the value of the button when is release.
 Can you help me?
Thanks

**Reply by tizfra** [reply]
So, this router rule `- {type: cc, chan: 1, par1: 21, par2: 11-20, patch: 1+}` works like a charm, it increment the patch one by one;
this one `- {type: cc, chan: 1, par1: 21, par2: 10, patch: 1}` doen't do nothing, and also looks like fluidpatcher doesnt receive the value that I send, midi monitor show me only the value 0 when I release the button I use to send `10` on midi cc 21.
I think there is a bug in router rules.
Is that true?
@albedozero tell me if you need more information about it.

**Reply by albedozero** [reply]
The MIDI monitor on the SquishBox should show all incoming MIDI messages, regardless of what the router rules are. Maybe your controller is doing something weird on button press? You can confirm that your controller is really sending a value of 10 on CC21 using `aseqdump` at the command line - explained [here](https://github.com/GeekFunkLabs/fluidpatcher/blob/master/troubleshooting.md#monitor-midi-messages). However, adding the `chan` specification to the rule shouldn't be needed - without it the rule should apply to messages from any channel. So, you could be right that there is a bug - trying to track it down now

**Reply by tizfra** [reply]
Yes, I'm sure that the controller sends the correct value on cc21, because if I delete `chan: 1` from the rule I can see the value on the midi monitor.

**Reply by tizfra** [reply]
I'm playing from fluidpatcher on Windows 10 notebook, but I think there isn't any difference between fluidpatcher on Windows and squishbox.
Is that right?

**Reply by albedozero** [reply]
I think #95 fixes this - good spot and thanks for pointing it out! I'll update the release so users can update via the install script

**Reply by albedozero** [reply]
> I'm playing from fluidpatcher on Windows 10 notebook, but I think there isn't any difference between fluidpatcher on Windows and squishbox. Is that right?

That's correct

**Reply by albedozero** [reply]
New release is [here](https://github.com/GeekFunkLabs/fluidpatcher/releases/tag/v0.8.3)

**Reply by tizfra** [reply]
Thanks for the new release, this rule `- {type: cc, par1: 21, par2: 1-10, patch: 1-}`  now works perfectly.
Can you help me to find where is the problem in this other router rule `- {type: cc, par1: 22, par2: 1, patch: 0}`?

When I send CC22=1, I want to select the patch 0, but with this rule I don't see 22=1 in midi monitor, I see all the other values, but not the 1.
I've also done the test with MIDI-OX and a virtual midi cable, to see that I send the value I want.

**Reply by albedozero** [reply, marked as answer]
This  was  a different but related bug, now fixed by #96 and updated in the latest release. Also, I decided that patch numbers should start with 1, rather than zero, so that they match what is shown on the squishbox display.

**Reply by tizfra** [reply]
Tested and now works perfectly.

Thanks for the help
